cpp/src/io/orc/stats_enc.cu

```diff
@@ -0,0 +1,380 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "orc_common.h"
+#include "orc_gpu.h"
+#include <io/utilities/block_utils.cuh>
+
+namespace cudf {
+namespace io {
+namespace orc {
+namespace gpu {
+
+/**
+ * @brief Initializes statistics groups
+ *
+ * @param[out] groups Statistics groups
+ * @param[in] cols Column descriptors
+ * @param[in] num_columns Number of columns
+ * @param[in] num_rowgroups Number of rowgroups
+ * @param[in] row_index_stride Rowgroup size in rows
+ *
+ **/
+// blockDim {128,1,1}
+__global__ void __launch_bounds__(128)
+gpuInitStatisticsGroups(statistics_group *groups, const stats_column_desc *cols,
+                        uint32_t num_columns, uint32_t num_rowgroups, uint32_t row_index_stride) {
+  __shared__ __align__(4) volatile statistics_group grp_g[4];
+  uint32_t col_id = blockIdx.y;
+  uint32_t ck_id = (blockIdx.x * 4) + (threadIdx.x >> 5);
+  uint32_t t = threadIdx.x & 0x1f;
+  volatile statistics_group *grp = &grp_g[threadIdx.x >> 5];
+  if (ck_id < num_rowgroups) {
+    if (!t) {
+      uint32_t num_rows = cols[col_id].num_rows;
+      grp->col = &cols[col_id];
+      grp->start_row = ck_id * row_index_stride;
+      grp->num_rows = min(num_rows - min(ck_id * row_index_stride, num_rows), row_index_stride);
+      __threadfence_block();
+    }
+    SYNCWARP();
+    if (t < sizeof(statistics_group) / sizeof(uint32_t)) {
+      reinterpret_cast<uint32_t *>(&groups[col_id * num_rowgroups + ck_id])[t] = reinterpret_cast<volatile uint32_t *>(grp)[t];
+    }
+  } 
+}
+
+
+/**
+ * @brief Get the buffer size and offsets of encoded statistics
+ *
+ * @param[in,out] groups Statistics merge groups
+ * @param[in] statistics_count Number of statistics buffers
+ *
+ **/
+// blockDim {1024,1,1}
+__global__ void __launch_bounds__(1024, 1)
+gpuInitStatisticsBufferSize(statistics_merge_group *groups, const statistics_chunk *chunks, uint32_t statistics_count) {
+  __shared__ volatile uint32_t scratch_red[32];
+  __shared__ volatile uint32_t stats_size;
+  uint32_t t = threadIdx.x;
+  if (!t) {
+    stats_size = 0;
+  }
+  __syncthreads();
+  for (uint32_t start = 0; start < statistics_count; start += 1024) {
```

Comment:
Use a constexpr variable for block size rather than hard-coding in multiple places.